proposals/p6333.md

```diff
@@ -0,0 +1,488 @@
+used, all prelude files and possibly more of the `Core` package will be
+re-checked, along with C++ ASTs being reproduced.
+
+Instead, Carbon could serialize checked IR, store produced C++ ASTs, and so on.
+C++ ASTs in particular could be substantially constructed based on parsed Carbon
+state, rather than checked Carbon state, allowing more build parallelism. In
+distributed or cached build systems, being able to reuse portions of the build
+may increase performance.
+
+The specific build outputs we want to store may substantially affect how we
+would set up a build process. The absence of a decision may lead to the
+implementation diverging from what's actually needed, meaning parts will be
+reimplemented later. This isn't expected to be too high cost.
+
+There are also ways to improve build performance without taking these steps.
+[Clang modules](https://clang.llvm.org/docs/Modules.html) might be used for
+improving Clang compile performance without significant support from Carbon.
+
+For now we will rely on whatever caching Bazel does for the `.a` output of a
+`carbon_library`. No other outputs will be made available. That may change, but
+leads want to spend our limited development and review time on other features
+for the 0.1 milestone.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   `carbon build` should support easy experimentation with Carbon, and also
+        small projects.
+    -   Other build support is intended to scale up for larger codebases.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   The intent is to be able to migrate a CMake, Makefile, or other build at
+        relatively low cost. An invocation to `clang` can typically be replaced
+        with `carbon clang`, linking a binary becomes `carbon link`, and so on.
+    -   Similarly, `carbon_library` and `carbon_binary` are important to us for
+        Bazel support and a migration from `cc_library` and `cc_binary`.
+
+## Alternatives considered
+
+### Naming of commands and rules
+
+For `carbon compile` and `carbon build`, this is trying to split apart concepts.
+Some considered alternatives are:
+
+-   Merge `compile`, and possibly also `link`, into `build`. Flags could be used
+    to differentiate between the versions desired, rather than subcommand names.
+    -   We expect that splitting these apart makes it easier to turn them into
+        replacements in C++ builds, and easier to understand even in
+        Carbon-specific builds.
+-   Have `carbon build` produce `a.out`
+    -   `a.out` is the default output of most C++ compilers, but it reflects a
+        legacy executable file format. Using the legacy name may reflect
+        backwards compatibility that Carbon doesn't plan.
+    -   Changing the default output name is probably low-cost, and people will
+        get used to it.
```

Comment:
What alternative is being proposed? I didn't find the alternative rule in the proposal -- are we picking a different name or are we requiring an explicit `-o` flag?